tree/ntuple/v7/src/RPageStorageDaos.cxx

```diff
@@ -0,0 +1,475 @@
+   }
+   auto zippedBytes = packedBytes;
+
+   if (fOptions.GetCompression() != 0) {
+      zippedBytes = fCompressor(buffer, packedBytes, fOptions.GetCompression());
+      if (!isAdoptedBuffer)
+         delete[] buffer;
+      buffer = const_cast<unsigned char *>(reinterpret_cast<const unsigned char *>(fCompressor.GetZipBuffer()));
+      isAdoptedBuffer = true;
+   }
+
+   auto offsetData = std::get<0>(fDaosContainer->WriteObject(buffer, zippedBytes,
+                                                         kDistributionKey, kAttributeKey)).lo;
+   fClusterMinOffset = std::min(offsetData, fClusterMinOffset);
+   fClusterMaxOffset = std::max(offsetData + zippedBytes, fClusterMaxOffset);
+
+   if (!isAdoptedBuffer)
+      delete[] buffer;
+
+   RClusterDescriptor::RLocator result;
+   result.fPosition = offsetData;
+   result.fBytesOnStorage = zippedBytes;
+   return result;
+}
+
+
+// TODO(jalopezg): the current byte range arithmetic makes little sense for the
+// object store. We might find out, however, that there are native ways to group
+// clusters in DAOS.
+ROOT::Experimental::RClusterDescriptor::RLocator
+ROOT::Experimental::Detail::RPageSinkDaos::CommitClusterImpl(ROOT::Experimental::NTupleSize_t /* nEntries */)
+{
+   RClusterDescriptor::RLocator result;
+   result.fPosition = fClusterMinOffset;
+   result.fBytesOnStorage = fClusterMaxOffset - fClusterMinOffset;
+   fClusterMinOffset = std::uint64_t(-1);
+   fClusterMaxOffset = 0;
+   return result;
+}
+
+
+void ROOT::Experimental::Detail::RPageSinkDaos::CommitDatasetImpl()
+{
+   const auto &descriptor = fDescriptorBuilder.GetDescriptor();
+   auto szFooter = descriptor.SerializeFooter(nullptr);
+   auto buffer = std::unique_ptr<unsigned char []>(new unsigned char[szFooter]);
+   descriptor.SerializeFooter(buffer.get());
+
+   auto zipBuffer = std::unique_ptr<unsigned char[]>(new unsigned char[szFooter]);
+   auto szZipFooter = fCompressor(buffer.get(), szFooter, fOptions.GetCompression(),
+      [&zipBuffer](const void *b, size_t n, size_t o){ memcpy(zipBuffer.get() + o, b, n); } );
+   WriteNTupleFooter(zipBuffer.get(), szZipFooter, szFooter);
+   WriteNTupleAnchor();
+}
+
+
+void ROOT::Experimental::Detail::RPageSinkDaos::WriteNTupleHeader(
+		const void *data, size_t nbytes, size_t lenHeader)
+{
+   fDaosContainer->WriteObject(kOidHeader, data, nbytes, kDistributionKey, kAttributeKey);
+   fNTupleAnchor.fLenHeader = lenHeader;
+   fNTupleAnchor.fNBytesHeader = nbytes;
+}
+
+void ROOT::Experimental::Detail::RPageSinkDaos::WriteNTupleFooter(
+		const void *data, size_t nbytes, size_t lenFooter)
+{
+   fDaosContainer->WriteObject(kOidFooter, data, nbytes, kDistributionKey, kAttributeKey);
+   fNTupleAnchor.fLenFooter = lenFooter;
+   fNTupleAnchor.fNBytesFooter = nbytes;
+}
+
+void ROOT::Experimental::Detail::RPageSinkDaos::WriteNTupleAnchor() {
+   fDaosContainer->WriteObject(kOidAnchor, &fNTupleAnchor, sizeof(fNTupleAnchor),
```

Comment:
This is not platform independent.  We need to actually serialize and deserialize the anchor members.

Reply:
True.  Thanks for spotting!